[ty] Support scoped declarations

iyakushev opened the pull request:
## Summary

Following ty issue [#698](https://github.com/astral-sh/ty/issues/698) this PR adds support for declarations.   
I'd say the implementation is rather naive, so I would use it as a starting point for a gradual improvement. I tried to check declarations before bindings and it seemed to resolved the issue. 

## Test Plan

Tested against mdtest (specifically attributes).

sharkdp commented:
Thank you for working on this! I am planning to take a closer look tomorrow.

It looks like this leads to a lot new false positives on ecosystem projects (see the auto-generated mypy_primer comment). We should probably try to find out what causes them (+ fix them and write a few more test cases here).

iyakushev commented:
Thank you @sharkdp! I will try to add some test cases following mypy_primer results tomorrow. Is it okay to add them into mdtests? 
As for false positives, I think its due to [this line](https://github.com/astral-sh/ruff/pull/19048/files#diff-a4d813e6b59314ddbc2448d1133969548760611a9de2d07a0c36a0cbe2c93ed3R1819). I mostly copied the way bindings did it while getting familiar with the codebase.

sharkdp commented on crates/ty_python_semantic/src/types/class.rs:
```diff
@@ -1748,6 +1748,46 @@ impl<'db> ClassLiteral<'db> {
+        for (attribute_declarations, method_scope_id) in
+            attribute_declarations(db, class_body_scope, name)
+        {
+            let method_scope = method_scope_id.to_scope_id(db, file);
+            if let Some(method_def) = method_scope.node(db).as_function(&module) {
+                let method_name = method_def.name.as_str();
+                if let Place::Type(Type::FunctionLiteral(method_type), _) =
+                    class_symbol(db, class_body_scope, method_name).place
+                {
+                    let method_decorator = MethodDecorator::try_from_fn_type(db, method_type);
+                    if method_decorator != Ok(target_method_decorator) {
+                        continue;
+                    }
+                }
+            }
```
Do you see a chance to avoid the code duplication here? This looks almost identical to what we have for bindings below.

iyakushev commented:
Thank you for taking time to review this! 
I've pushed the changes addressing your comments. As for duplication, I've moved the code into a closure, if that's ok. The name might be a little off imho.

sharkdp reviewed (approved):
This looks great — thank you very much!
